### ISSUE TITLE
Transform json_agg to a pREST native function

### ISSUE THREAD

**Issue by vmesel** [prest/prest#730]
Hey guys,

I'm using pREST on my development environment and I was getting out of memory error due to our `SELECT json_agg(s) FROM (<query here>) s`. This transformation is very costly for the bank (at least for the size of table I'm working with), and I'm migrating the implementation of QueryCtx (postgres.go:605) to a neutral approach.

Here is the code that is working partially:

```go
func (adapter *Postgres) QueryCtx(ctx context.Context, SQL string, params ...interface{}) (sc adapters.Scanner) {
	// use the db_name that was set on request to avoid runtime collisions
	db, err := getDBFromCtx(ctx)
	if err != nil {
		log.Errorln(err)
		return &scanner.PrestScanner{Error: err}
	}
	log.Debugln("generated SQL:", SQL, " parameters: ", params)
	p, err := Prepare(db, SQL)
	if err != nil {
		log.Errorln(err)
		return &scanner.PrestScanner{Error: err}
	}
	var jsonData []byte
	rows, err := p.QueryContext(ctx, params...)
	if err != nil {
		log.Fatal(err)
	}
	var data []map[string]interface{}
	for rows.Next() {
		cols, err := rows.Columns()
		if err != nil {
			log.Fatal(err)
		}
		columns := make([]interface{}, len(cols))
		columnPointers := make([]interface{}, len(cols))
		for i := range columns {
			columnPointers[i] = &columns[i]
		}
		if err := rows.Scan(columnPointers...); err != nil {
			log.Fatal(err)
		}
		m := make(map[string]interface{})
		for i, colName := range cols {
			val := columnPointers[i].(*interface{})
			switch (*val).(type) {
				case []uint8:
					var f interface{}
					err = json.Unmarshal((*val).([]byte), &f)
					m[colName] = f
				default:
					m[colName] = *val
			}
		}
		data = append(data, m)
	}
	jsonData, _ = json.Marshal(data)
	if len(jsonData) == 0 {
		jsonData = []byte("[]")
	}
	return &scanner.PrestScanner{
		Error:   err,
		Buff:    bytes.NewBuffer(jsonData),
		IsQuery: true,
	}
}```

Error got from Datagrip:
<img width="1075" alt="image" src="https://user-images.githubusercontent.com/4984147/199071959-2a64a979-923e-4ed0-a24a-c13105be295d.png">

**Comment by avelino** [comment]
@vmesel [this discussion is worth reading](https://github.com/prest/prest/discussions/387#discussioncomment-108952).

The premise of prestd is to be specialized on postgres databases and `json_agg` has existed since [postgres version **9.3**](https://www.postgresql.org/docs/9.3/release-9-3-6.html), I don't know if it makes sense to bring this processing into prestd, instead of outsourcing the responsibility to postgres.

cc @fabriziomello

**Comment by vmesel** [comment]
@avelino I don't think the discussion of the name of the tool must include its own features. I might be misjudging the matter, but the problem is: we are sending one new operation to the database, that depending on the size of the query, lacks memory of the database.

If I run the query as it is, the database system works just fine, but if I put it inside `SELECT json_agg(s) FROM (<MY AWESOME QUERY HERE>) s`, the database shows me the lack of memory message.

I sincerely think that all of the data manipulation outside of the users' scope must be done on the code (except for very specific and nitty SQL details and syntactic sugars that we might add)

**Comment by fabriziomello** [comment]
@vmesel @avelino the problem is that the `json_agg` function rely on `stringinfo.c` implementation that have some memory limitation: https://github.com/postgres/postgres/blob/master/src/common/stringinfo.c#L28.

@vmesel did u tried to just change the `json_agg` to `jsonb_agg` in the pREST source code???

**Comment by vmesel** [comment]
@fabriziomello no, I'm going to try it out.

**Comment by vmesel** [comment]
@fabriziomello tested it out: It doesn't explode the memory usage, **but** it changes the error to: `context deadline exceeded`.

<img width="444" alt="image" src="https://user-images.githubusercontent.com/4984147/199085288-e0314118-e80d-45ec-9f26-eacc2617bbc4.png">

**Comment by avelino** [comment]
@vmesel the question is: will this use be a real case for prestd?

We will probably have a similar problem with the `Go` implementation, remembering that we are running over an http protocol that has a timeout.

**Comment by fabriziomello** [comment]
And the @avelino points are good... change where the aggregation occurs doesn't means that this problem is solved.

**Comment by vmesel** [comment]
> @vmesel what is the size of the table you're trying to transform into a big JSON document?

22 GB.

**Comment by fabriziomello** [comment]
> > @vmesel what is the size of the table you're trying to transform into a big JSON document?
> 
> 22 GB.

Wow... do u really need such a big json document???

**Comment by vmesel** [comment]
Maybe I'm too crazy and might be pushing an unnecessary issue, if this is the case, I'm sorry guys.

**Comment by vmesel** [comment]
@fabriziomello the thing is, the json_agg is from pREST not from my query.

**Comment by fabriziomello** [comment]
@vmesel Yep... I understood it... what about trying the  `SELECT jsonb_agg(s) FROM (<your query>) s` connected direct to the Postgres database using `psql` client tool (not GUI tools that can do crazy things to get data from server to client)

**Comment by fabriziomello** [comment]
@vmesel before run the query execute this command in `psql` to get some more detailed information `\set VERBOSITY verbose`

**Comment by fabriziomello** [comment]
@vmesel seems the `jsonb_agg` test lead to OOM Killer take an action. Does your Postgres instance (and Operating System) properly tuned to deal with so large amount of data???

**Comment by fabriziomello** [comment]
We can improve the way that we build the JSON document, but to me it should be something like:
1. instead of execute the query in one shoot create a cursor
2. fetch data in batches (configurable size?)
3. append the batched to a local `jsonData` variable in pREST

**Comment by vmesel** [comment]
@fabriziomello when we batch, it works, but it uses more memory than necessary. I'm newbie to Golang, so I don't know how to set up a cursor, I'll try to find it and come back here if we all agree this is the viable way of making progress.

**Comment by arxdsilva** [comment]
@vmesel getting all info at once or too big will either one rate your DB or prestd app.

The question is: do you really need it at once? 

This is why: Most large scale apps have millions to billions of lines but you won't find one that loads it all, they do the loading as needed by the client(s).

So could you tell us your specific case?

**Comment by vmesel** [comment]
I can't tell a lot, it's a confidential use case for now, unfortunately.

But I've run `EXPLAIN ANALYZE` on both statements (both using offset and limit, one using json_agg and the other not using it).

```sql
explain analyze SELECT * FROM "database"."public"."table" LIMIT 50 OFFSET(1 - 1) * 50
```

Result:

```
Limit  (cost=0.00..5.29 rows=50 width=1285) (actual time=0.011..0.026 rows=50 loops=1)
  ->  Seq Scan on mv_screener_v3_test_vm  (cost=0.00..3150542.78 rows=29758878 width=1285) (actual time=0.011..0.021 rows=50 loops=1)
Planning Time: 0.088 ms
Execution Time: 0.054 ms
```

The one with the `json_agg` statement:

```
explain analyze SELECT json_agg(s) FROM (SELECT * FROM "database"."public"."table" LIMIT 50 OFFSET(1 - 1) * 50) s
```

Result:

```
Aggregate  (cost=5.92..5.93 rows=1 width=32) (actual time=1.440..1.440 rows=1 loops=1)
  ->  Subquery Scan on s  (cost=0.00..5.79 rows=50 width=1309) (actual time=0.024..0.136 rows=50 loops=1)
        ->  Limit  (cost=0.00..5.29 rows=50 width=1285) (actual time=0.011..0.030 rows=50 loops=1)
              ->  Seq Scan on mv_screener_v3_test_vm  (cost=0.00..3150542.78 rows=29758878 width=1285) (actual time=0.011..0.024 rows=50 loops=1)
Planning Time: 0.147 ms
Execution Time: 1.480 ms
```

The execution time with the json_agg is 27x less efficient than when running the normal query.

Using the code from PR #731, the request go as:

```

2022/11/03 11:10:51 [warning] adapter is not set. Using the default (postgres)
2022/11/03 11:10:53 [warning] You are running prestd in public mode.
2022/11/03 11:10:53 [warning] command.go:920 You are running prestd in debug mode.
[prestd] listening on 0.0.0.0:8080 and serving on /
2022/11/03 11:11:15 [debug] server.go:2109 generated SQL:SELECT * FROM "database"."public"."table" LIMIT 100 OFFSET(1 - 1) * 100 parameters: []
[negroni] 2022-11-03T11:11:15-03:00 | 200 | 	 12.631549357s | localhost:8080 | GET /database/public/table
[negroni] 2022-11-03T11:11:28-03:00 | 404 | 	 169.6µs | localhost:8080 | GET /favicon.ico
```

And the code with the json_agg (commit: ba97cfd)

```
2022/11/03 11:12:27 [warning] adapter is not set. Using the default (postgres)
2022/11/03 11:12:29 [warning] You are running prestd in public mode.
2022/11/03 11:12:29 [warning] command.go:920 You are running prestd in debug mode.
[prestd] listening on 0.0.0.0:8080 and serving on /
2022/11/03 11:14:55 [debug] server.go:2109 generated SQL:SELECT json_agg(s) FROM (SELECT * FROM "database"."public"."table" LIMIT 100 OFFSET(1 - 1) * 100) s parameters: []
[negroni] 2022-11-03T11:14:55-03:00 | 200 | 	 29.619207355s | localhost:8080 | GET /database/public/table
```

As we can see, `json_agg` on the database side takes twice the time as the code available in #731. It would be more efficient and, IMHO, more intelligent to serialize data on the web server side than on the database.

Does this help you @arxdsilva ?

**Comment by avelino** [comment]
Does using `jsonb_agg` improve it?

**Comment by vmesel** [comment]
@avelino not at all, actually it decreases efficiency by 36 times.

```
Aggregate  (cost=5.92..5.93 rows=1 width=32) (actual time=36.150..36.151 rows=1 loops=1)
  ->  Subquery Scan on s  (cost=0.00..5.79 rows=50 width=1309) (actual time=0.029..32.193 rows=50 loops=1)
        ->  Limit  (cost=0.00..5.29 rows=50 width=1285) (actual time=0.016..0.050 rows=50 loops=1)
              ->  Seq Scan on table  (cost=0.00..3150542.78 rows=29758878 width=1285) (actual time=0.016..0.043 rows=50 loops=1)
Planning Time: 0.169 ms
Execution Time: 36.317 ms
```

**Comment by avelino** [comment]
**OMG**

This will absurdly improve query processing time.
Looking at this data I think it is worth investing in the implementation of Go serialization (even though I am afraid of breaking something, we need to write a lot of tests).

Thinking about the release it is important to have a chart showing the performance improvement comparing the 3 versions.

**Comment by vmesel** [comment]
So, @avelino the PoC is already half-way through! Take a look at #731. We just need to check and implement tests. I'm with some doubts regarding typing and some other things but first, look at the code. The doubts I'll send you them through Telegram as they are not about the project in itself.

**Comment by arxdsilva** [comment]
> As we can see, `json_agg` on the database side takes twice the time as the code available in #731. It is more efficient and, IMHO, more intelligent to serialize data on the web server side than on the database.
> 
> Does this help you @arxdsilva ?

Thanks for clarifying @vmesel , could you also do the same query but using the high level API on both cases?

I ask this because this can be a big change for some users and I fear that just by removing the processing from DB and moving to the app may not have all this performance advantage as either way you'd need to to all this json transformation in the app anyways.

**Comment by vmesel** [comment]
@arxdsilva is it different than this: https://github.com/prest/prest/issues/730#issuecomment-1302183157 ?

**Comment by vmesel** [comment]
With my changes

```
λ ~/Github/prest/ fix/transform-query-to-json go run cmd/prestd/main.go

2022/11/04 11:24:11 [warning] adapter is not set. Using the default (postgres)
2022/11/04 11:24:13 [warning] You are running prestd in public mode.
2022/11/04 11:24:13 [warning] command.go:920 You are running prestd in debug mode.
[prestd] listening on 0.0.0.0:8080 and serving on /
2022/11/04 11:24:13 [debug] server.go:2109 generated SQL:SELECT * FROM "database"."public"."table" LIMIT 100 OFFSET(1 - 1) * 100 parameters: []
[negroni] 2022-11-04T11:24:13-03:00 | 200 | 	 3.818255218s | localhost:8080 | GET /database/public/table
2022/11/04 11:24:24 [debug] server.go:2109 generated SQL:SELECT * FROM "database"."public"."table" LIMIT 1000 OFFSET(1 - 1) * 1000 parameters: []
[negroni] 2022-11-04T11:24:24-03:00 | 200 | 	 5.122278833s | localhost:8080 | GET /database/public/table
```

With JSON_AGG():

```
λ ~/Github/prest/ main go run cmd/prestd/main.go


2022/11/04 11:25:26 [warning] adapter is not set. Using the default (postgres)
2022/11/04 11:25:28 [warning] You are running prestd in public mode.
2022/11/04 11:25:28 [warning] command.go:920 You are running prestd in debug mode.
[prestd] listening on 0.0.0.0:8080 and serving on /
2022/11/04 11:25:28 [debug] server.go:2109 generated SQL:SELECT json_agg(s) FROM (SELECT * FROM "database"."public"."table" LIMIT 100 OFFSET(1 - 1) * 100) s parameters: []
[negroni] 2022-11-04T11:25:28-03:00 | 200 | 	 5.007617458s | localhost:8080 | GET /database/public/table
2022/11/04 11:25:46 [debug] server.go:2109 generated SQL:SELECT json_agg(s) FROM (SELECT * FROM "database"."public"."table" LIMIT 1000 OFFSET(1 - 1) * 1000) s parameters: []
[negroni] 2022-11-04T11:25:46-03:00 | 200 | 	 7.014840878s | localhost:8080 | GET /database/public/table
```

With JSONB_AGG():

```
2022/11/04 11:26:44 [warning] adapter is not set. Using the default (postgres)
2022/11/04 11:26:47 [warning] You are running prestd in public mode.
2022/11/04 11:26:47 [warning] command.go:920 You are running prestd in debug mode.
[prestd] listening on 0.0.0.0:8080 and serving on /
2022/11/04 11:26:49 [debug] server.go:2109 generated SQL:SELECT jsonb_agg(s) FROM (SELECT * FROM "database"."public"."table" LIMIT 100 OFFSET(1 - 1) * 100) s parameters: []
[negroni] 2022-11-04T11:26:49-03:00 | 200 | 	 3.765227403s | localhost:8080 | GET /database/public/table
2022/11/04 11:26:58 [debug] server.go:2109 generated SQL:SELECT jsonb_agg(s) FROM (SELECT * FROM "database"."public"."table" LIMIT 1000 OFFSET(1 - 1) * 1000) s parameters: []
[negroni] 2022-11-04T11:26:58-03:00 | 200 | 	 5.63465962s | localhost:8080 | GET /database/public/table
```

**Comment by vmesel** [comment]
Just as a matter of fact, I've created a t3.medium on AWS (2 vCPUs and 4GBs of RAM) on the same VPC as the database. @arxdsilva and @avelino what are your thoughts?

With JSON_AGG()

```
2022/11/04 16:54:35 [warning] adapter is not set. Using the default (postgres)
2022/11/04 16:54:35 [warning] command.go:920 You are running prestd in debug mode.
[prestd] listening on 0.0.0.0:80 and serving on /
2022/11/04 16:54:41 [debug] server.go:2084 generated SQL:SELECT json_agg(s) FROM (SELECT * FROM "database"."public"."table" LIMIT 1000 OFFSET(1 - 1) * 1000) s parameters: []
[negroni] 2022-11-04T16:54:41Z | 200 | 	 5.11392325s | 54.186.223.54 | GET /database/public/table
2022/11/04 16:55:18 [debug] server.go:2084 generated SQL:SELECT json_agg(s) FROM (SELECT * FROM "database"."public"."table" LIMIT 100 OFFSET(1 - 1) * 100) s parameters: []
[negroni] 2022-11-04T16:55:18Z | 200 | 	 504.78063ms | 54.186.223.54 | GET /database/public/table
```


With JSONB_AGG()
```
[prestd] listening on 0.0.0.0:80 and serving on /
2022/11/04 16:59:26 [debug] server.go:2084 generated SQL:SELECT jsonb_agg(s) FROM (SELECT * FROM "database"."public"."table" LIMIT 100 OFFSET(1 - 1) * 100) s parameters: []
[negroni] 2022-11-04T16:59:26Z | 200 | 	 479.260256ms | 54.186.223.54 | GET /database/public/table
2022/11/04 17:00:05 [debug] server.go:2084 generated SQL:SELECT jsonb_agg(s) FROM (SELECT * FROM "database"."public"."table" LIMIT 1000 OFFSET(1 - 1) * 1000) s parameters: []
[negroni] 2022-11-04T17:00:05Z | 200 | 	 1.912713761s | 54.186.223.54 | GET /database/public/table
```

With the aggregation function in our codebase:
```
2022/11/04 16:58:12 [warning] adapter is not set. Using the default (postgres)
2022/11/04 16:58:12 [warning] command.go:920 You are running prestd in debug mode.
[prestd] listening on 0.0.0.0:80 and serving on /
2022/11/04 16:58:23 [debug] server.go:2084 generated SQL:SELECT * FROM "database"."public"."table" LIMIT 1000 OFFSET(1 - 1) * 1000 parameters: []
[negroni] 2022-11-04T16:58:23Z | 200 | 	 1.861196314s | 54.186.223.54 | GET /database/public/table
2022/11/04 16:58:42 [debug] server.go:2084 generated SQL:SELECT * FROM "database"."public"."table" LIMIT 100 OFFSET(1 - 1) * 100 parameters: []
[negroni] 2022-11-04T16:58:42Z | 200 | 	 32.047049ms | 54.186.223.54 | GET /database/public/table
```

**Comment by vmesel** [comment]
Thanks for the edit @avelino!

**Comment by vmesel** [comment]
As I discussed with @avelino through DMs, we arrived at a consensus: If we change the `json_agg` to `jsonb_agg` we can improve the response time and focus our effort on new features and fixing bugs on our platform.

PR with JSONB_AGG change available on #733
